File: zig/pkg/antfly/src/storage/db/enrichment/chunker.zig
// Copyright 2026 Antfly, Inc.
//
// Licensed under the Elastic License 2.0 (ELv2); you may not use this file
// except in compliance with the Elastic License 2.0. You may obtain a copy of
// the Elastic License 2.0 at
//
//     https://www.antfly.io/licensing/ELv2-license
//
// Unless required by applicable law or agreed to in writing, software distributed
// under the Elastic License 2.0 is distributed on an "AS IS" BASIS, WITHOUT
// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
// Elastic License 2.0 for the specific language governing permissions and
// limitations.

const std = @import("std");
const Allocator = std.mem.Allocator;
const chunking = @import("../../../chunking/mod.zig");

pub const Chunk = chunking.chunk.Chunk;

pub fn chunkText(
    alloc: Allocator,
    text: []const u8,
    chunk_size: u32,
    chunk_overlap: u32,
) ![]Chunk {
    if (chunk_size == 0) return &.{};
    if (chunk_overlap >= chunk_size) return error.InvalidChunkOverlap;

    var chunks = std.ArrayListUnmanaged(Chunk).empty;
    errdefer {
        for (chunks.items) |*chunk| chunk.deinit(alloc);
        chunks.deinit(alloc);
    }

    var chunk_id: u32 = 0;
    const step = chunk_size - chunk_overlap;
    var start: usize = 0;
    while (start < text.len) : (start += step) {
        const end = @min(start + chunk_size, text.len);
        try chunks.append(alloc, .{
            .chunk_id = chunk_id,
            .text = try alloc.dupe(u8, text[start..end]),
            .start_offset = std.math.cast(u32, start),
            .end_offset = std.math.cast(u32, end),
        });
        chunk_id += 1;
        if (end == text.len) break;
    }

    return try chunks.toOwnedSlice(alloc);
}

pub fn chunkTextWithConfigJson(
    alloc: Allocator,
    text: []const u8,
    config_json: []const u8,
) ![]Chunk {
    var cfg = try chunking.types.parseConfigFromSlice(alloc, config_json);
    defer cfg.deinit(alloc);

    return switch (cfg.provider) {
        .mock => try chunking.fixed.chunkText(alloc, text, cfg),
        .antfly => try chunking.inference.chunkText(alloc, cfg, text),
    };
}

pub fn freeChunks(alloc: Allocator, chunks: []Chunk) void {
    for (chunks) |*chunk| chunk.deinit(alloc);
    alloc.free(chunks);
}

test "chunker splits overlapping windows" {
    const alloc = std.testing.allocator;

    const chunks = try chunkText(alloc, "abcdefghij", 4, 1);
    defer freeChunks(alloc, chunks);

    try std.testing.expectEqual(@as(usize, 3), chunks.len);
    try std.testing.expectEqualStrings("abcd", chunks[0].text.?);
    try std.testing.expectEqualStrings("defg", chunks[1].text.?);
    try std.testing.expectEqualStrings("ghij", chunks[2].text.?);
    try std.testing.expectEqual(@as(?u32, 0), chunks[0].start_offset);
    try std.testing.expectEqual(@as(?u32, 4), chunks[0].end_offset);
    try std.testing.expectEqual(@as(?u32, 3), chunks[1].start_offset);
}

File: go/pkg/antfly/lib/chunking/chunker.go
// Copyright 2026 Antfly, Inc.
//
// Licensed under the Elastic License 2.0 (ELv2); you may not use this file
// except in compliance with the Elastic License 2.0. You may obtain a copy of
// the Elastic License 2.0 at
//
//     https://www.antfly.io/licensing/ELv2-license
//
// Unless required by applicable law or agreed to in writing, software distributed
// under the Elastic License 2.0 is distributed on an "AS IS" BASIS, WITHOUT
// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
// Elastic License 2.0 for the specific language governing permissions and
// limitations.

//go:generate go tool oapi-codegen --config=cfg.yaml ../../../../../specs/openapi/antfly/chunking.yaml
package chunking

import (
	"context"
	"fmt"

	externalRef0 "github.com/antflydb/antfly/go/pkg/libaf/chunking"
)

// NewTextChunk creates a Chunk containing text content with the given parameters.
// Re-exported from libaf/chunking for convenience.
var NewTextChunk = externalRef0.NewTextChunk

// TextChunkOptions is re-exported from libaf/chunking for convenience.
type TextChunkOptions = externalRef0.TextChunkOptions

// AudioChunkOptions is re-exported from libaf/chunking for convenience.
type AudioChunkOptions = externalRef0.AudioChunkOptions

// Chunker splits text into semantically meaningful chunks
type Chunker interface {
	Chunk(ctx context.Context, text string) ([]Chunk, error)
	ChunkMedia(ctx context.Context, data []byte, mimeType string) ([]Chunk, error)
	Close() error
}

// ChunkerRegistry maps provider names to constructor functions
var ChunkerRegistry = map[ChunkerProvider]func(config ChunkerConfig) (Chunker, error){}

// NewChunker creates a new chunker instance based on the provider configuration
func NewChunker(config ChunkerConfig) (Chunker, error) {
	if config.Provider == "" {
		return nil, fmt.Errorf("provider is required")
	}

	constructor, ok := ChunkerRegistry[config.Provider]
	if !ok {
		return nil, fmt.Errorf("unknown chunker provider: %s", config.Provider)
	}

	return constructor(config)
}

// defaultChunkerConfig is the default chunker configuration, set from config at startup.
var defaultChunkerConfig *ChunkerConfig

// SetDefaultChunkerConfig sets the default chunker configuration.
// This should be called during config initialization.
func SetDefaultChunkerConfig(config *ChunkerConfig) {
	defaultChunkerConfig = config
}

// GetDefaultChunkerConfig returns the current default chunker configuration.
func GetDefaultChunkerConfig() *ChunkerConfig {
	return defaultChunkerConfig
}
